// @embroider/core@2.1.1/src/template-compiler-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeTemplateCompiler = void 0;
const path_1 = require("path");
const load_ember_template_compiler_1 = require("./load-ember-template-compiler");
const template_compiler_common_1 = require("./template-compiler-common");
class NodeTemplateCompiler extends template_compiler_common_1.TemplateCompiler {
    constructor(params) {
        super({
            loadEmberTemplateCompiler: () => (0, load_ember_template_compiler_1.getEmberExports)(params.compilerPath),
            resolver: params.resolver,
            EmberENV: params.EmberENV,
            plugins: params.plugins,
        });
        this.params = params;
    }
    // Use applyTransforms on the contents of inline hbs template strings inside
    // Javascript.
    inlineTransformsBabelPlugin() {
        return [
            (0, path_1.join)(__dirname, 'babel-plugin-stage1-inline-hbs-node.js'),
            {
                templateCompiler: this.params,
            },
        ];
    }
    baseDir() {
        return (0, path_1.join)(__dirname, '..');
    }
}
exports.NodeTemplateCompiler = NodeTemplateCompiler;
